components/scream/src/share/field/field.hpp

```diff
@@ -93,6 +117,9 @@ class Field {
 
   // Keep track of whether the field has been allocated
   bool                            m_allocated;
+
+  // List of property checks for this field.
+  property_check_list m_prop_checks;
```

Comment:
I think we should put this in the FieldHeader. The header is "shared" across all copies of the Field, so that if someone adds a property check in the header, all other copies of Field will see it. On the other hands, if some atm proc adds a prop check to the field itself, all other copies of the field won't be affected.

Reply:
That sounds good. I'll move this and `add_property_check` to the FieldHeader class. Do you think we should keep the `property_check_begin()` and `property_check_end()` iterators on the Field class, or move them to the FieldHeader as well?

Reply:
There's an issue with this that introduces the usual inheritance-related C++ warts. :-( FieldPropertyCheck is a template class with `ScalarType` and `Device` parameters, but `FieldHeader` is just a regular class. So now we have to make an otherwise-pointless `FieldPropertyCheckBase` class. :-( This is an anti-pattern, as far as I'm concerned, because it introduces issues that we have to understand on a per-case basis (such as: what if we want to check a property on fields having different scalar types and/or resident devices?).

If we can assume that properties are checked in a specific context (say, by the FieldRepo, in which we know our ScalarType is Real), we can simplify things a bit. Do you have any thoughts about this?

Reply:
Ah, that's a good point. I have no strong preference. That said, I think we're mixing a bit the templated and non-templated approaches in scream: fields are templated on scalar/device, but atm processes hard code a fixed one everywhere.

I think at some point we will _have_ to go toward a `FieldBase` class, which defines _some_ interfaces, but not all (in particular, no interface to get the actual view). I've seen the patter in other libraries, and albeit not great, it can be done.

For now, I would say leave FPC inside the Field class, so we can get the concept merged in. We'll have to go back on the field concept soon, and possibly change design (e.g., with a FieldBase class).

Reply:
I think a compromise to this issue is to store a shared_ptr<property_check_list>. This way, if AtmProc1 adds a property check, it will be visible also inside AtmProc2.

I am not 100% sure this is ever going to be used (meaning, one proc adds a check that another proc will also want to run), but I can't exclude it. And the fix (turning the list into a ptr to a list) is probably small enough to just do it.

Reply:
Sounds good!

Reply:
Actually, I think we _do_ need this. In particular, we may have the checks done outside the atm proc (otherwise why storing them in the field?), in which case, we need to make sure that if AtmProcX adds a prop test, it will be stored not only in its local copy, but also in all the copies of that field.